include/graphblas/reference/blas3.hpp

```diff
@@ -916,6 +916,206 @@ namespace grb {
 		return ret;
 	}
 
+
+	/**
+	 * A masked outer product of two vectors. Assuming vectors \a u and \a v are oriented
+	 * column-wise, the result matrix \a A will contain \f$ uv^T \f$, masked to non-zero values from \a mask.
+	 */
+	template<
+		Descriptor descr = descriptors::no_operation,
+		class Operator,
+		typename InputType1, typename InputType2, typename OutputType,
+		typename Coords,
+		typename RIT, typename CIT, typename NIT
+	>
+	RC maskedOuter(
+		Matrix< OutputType, reference, RIT, CIT, NIT > &A,
+		const Matrix< OutputType, reference, RIT, CIT, NIT > &mask,
+		const Vector< InputType1, reference, Coords > &u,
+		const Vector< InputType2, reference, Coords > &v,
+		const Operator &mul = Operator(),
+		const Phase &phase = EXECUTE,
+		const typename std::enable_if<
+			grb::is_operator< Operator >::value &&
+			!grb::is_object< InputType1 >::value &&
+			!grb::is_object< InputType2 >::value &&
+			!grb::is_object< OutputType >::value,
+			void >::type * const = nullptr
+	) {
+		// static checks
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D1, InputType1 >::value
+			), "grb::maskedOuter",
+			"called with a prefactor vector that does not match the first domain "
+			"of the given multiplication operator" );
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D2, InputType2 >::value
+			), "grb::maskedOuter",
+			"called with a postfactor vector that does not match the first domain "
+			"of the given multiplication operator" );
+		NO_CAST_ASSERT( ( !(descr & descriptors::no_casting) ||
+				std::is_same< typename Operator::D3, OutputType >::value
+			), "grb::maskedOuter",
+			"called with an output matrix that does not match the output domain of "
+			"the given multiplication operator" );
+		static_assert( ( !(descr & descriptors::invert_mask)
+			),
+			"grb::maskedOuter: invert_mask descriptor cannot be used ");
+#ifdef _DEBUG
+		std::cout << "In grb::maskedOuter (reference)\n";
+#endif
+
+		const size_t nrows = size( u );
+		const size_t ncols = size( v );
+
+		const size_t m = grb::nrows( mask );
+		const size_t n = grb::ncols( mask );
+
+		assert( phase != TRY );
+		if( nrows != grb::nrows( A ) || nrows != m ) {
+			return MISMATCH;
+		}
+
+		if( ncols != grb::ncols( A ) || ncols != n ) {
+			return MISMATCH;
+		}
+
```

Comment:
Another clause guard for *empty mask* or *empty u* or *empty v* would be nice

Reply:
Shouldn't the matrix just be empty in that case?

Reply:
Yes it should be cleared

Reply:
No wait there is a catch, if you read extremely attentively the doc: "empty mask" means "no mask". Hence, you should rely on the non-masked variant.

Reply:
Wait, I interpreted "empty mask" as it has nnz = 0? Is this the correct interpretation?

Reply:
Yes, "empty mask" means `nnz(mask)==0`

Reply:
Ok, did that, thanks for this case!

Reply:
@anyzelman Can you confirm ?

Reply:
Sorry, saw this late -- empty mask means `size(mask)==0`

Reply:
@aleksamilisavljevic Ok sorry for the confusion it's not `nnz(mask)==0` but `size(mask)==0`

Reply:
Resolved, thanks.